src/persist/src/indexed/runtime.rs

```diff
@@ -705,17 +792,51 @@ impl<L: Log, B: Blob> RuntimeImpl<L, B> {
             .cmd_run_ms
             .inc_by(metric_duration_ms(step_start.duration_since(run_start)));
 
-        if let Err(e) = self.indexed.step() {
-            self.metrics.cmd_step_error_count.inc();
-            // TODO: revisit whether we need to move this to a different log level
-            // depending on how spammy it ends up being. Alternatively, we
-            // may want to rate-limit our logging here.
-            log::warn!("error running step: {:?}", e);
-        }
+        // HACK: This rate limits how much we call step in response to workloads
+        // consisting entirely of write, seal, and allow_compaction (which is
+        // exactly what we expect in the common/steady state). At the moment,
+        // step is too aggressive about compaction if called in a tight loop,
+        // causing excess disk usage. This is exactly what happens if coord is
+        // getting steady read traffic over indexes derived directly or
+        // indirectly from tables (it continually seals).
+        //
+        // The downside to this rate limiting is that it introduces artificial
+        // latency into filling the response futures we returned to the client
+        // as well as to updating listeners. The former blocks literally no
+        // critical paths in the initial persisted system tables test, which is
+        // considered an acceptable tradeoff for the short-term. In the
+        // long-term, we'll do something better (probably by using Buffer).
+        //
+        // TODO: It would almost certainly be better to separate the rate limit
+        // of drain_pending vs the rest of step. The other parts (drain_unsealed
+        // and compact) can and should be called even less frequently than this.
+        // However, this change is going into a release at the last minute and
+        // I'm less confidant about unknown ramifications of a two interval
+        // strategy.
+        let need_step = step_start.duration_since(self.prev_step) > self.min_step_interval;
+
+        // BONUS HACK: If pending_responses is empty, then we know nothing has
+        // changed and step would be a no-op. Unfortunately, step is not yet
+        // smart about actually being a no-op and will still write out meta
+        // (three times, in fact).
+        //
+        // TODO: Make step smarter and remove this hack.
+        let need_step = need_step && !self.indexed.pending_responses.is_empty();
```

Comment:
So we would enter this state iff the only command we received was a `Tick` under normal operations (where we see sequences of `Write`, `Seal`, `AllowCompaction`, and `Tick` in steady state right?)

I'm trying to wrap my head around why `need_step` would be true, and `pending_responses` would be empty. I don't have a good intuition for how frequently `coord` calls `seal` or `allow_compaction` anymore and my sense is that for `mz_metrics` coord calls `write` every 30 seconds.

But if I'm understanding everything correctly it seems like -- let's say the tick interval is 1s, then we'll send ticks every 100 ms according to the code above, but then if we see `seal`s every 10 ms, say because someone is hitting mz with 100 selects a second (dunno if that's even possible) then most of the time the bonus hack won't help us at all (because we'll have pending seals to respond to every time we're actually allowed to call `step`)

Sorry if I'm missing something simple here, but its hard for me to see concretely what workloads the bonus hack is actually useful in.

Reply:
I guess the bonus hack is useful in cases when seals / allow compactions and writes are all performed less frequently than the step interval. in those cases, tick might tell us "hey you need to call step its been one second" but if we havent received any seals, allow compactions or writes in the past second, we can ignore the tick. Is that right?

Reply:
Yes, you got in exactly right in the followup. Added a comment. Fwiw, this optimization applies to all of our current release load tests :)

> I don't have a good intuition for how frequently coord calls seal or allow_compaction anymore and my sense is that for mz_metrics coord calls write every 30 seconds.

Added a comment with this info